### ISSUE TITLE
question about training code

### ISSUE THREAD

**Issue by jinxxo-j** [phatli/CaseVPR#3]
As you suggested, I have ported the HVPR_CaseNet model from the codebase into the vg-transformer code and am currently training it.
In the code you provided, the model is defined as follows:

```python
class HVPR_CaseNet(nn.Module):
    def __init__(self, seq_len=5, encoder_type="bs_d_c", is_pure=False):
        super().__init__()
        if is_pure:
            self.encoder = get_pure_dinov2(num_unfrozen_blocks=2)
        else:
            self.encoder = vit_base(patch_size=14, img_size=518,
                                    init_values=1, block_chunks=0)
        self.pool = CaseNet(seq_len=seq_len, encode_type=encoder_type)
        self.seq_length = seq_len

    def forward(self, x, single_img=False):
        features = self.encoder(x, is_training=True)
        img_desc, seq_desc = self.pool(features, output_img=True, single_img=single_img)
        return img_desc, seq_desc
``` 

However, since img_desc does not seem to be required for training, I modified the forward function as follows:

```python
def forward(self, x, single_img=False):
    features = self.encoder(x, is_training=True)
    seq_desc = self.pool(features, output_img=False, single_img=single_img)
    return seq_desc
``` 

Would this be the correct way to implement the modification?(output_img=False, single_img=False)

**Comment by phatli** [comment]
Yes, it looks correct.

**Comment by jinxxo-j** [comment]
Is there any additional training setting to reproduce CASEVPR?

I use vg-transformer's base setting and casevpr's setting.

I used oxford validation set. and early stop as vg-transformers used.



CUDA_VISIBLE_DEVICES=4 python main_train.py \
	--dataset_path /mnt/ssd4/jinwoo/data/robotcar_cut_2m \
	--img_shape 224 224 \
	--arch hvpr --aggregation casenet \
	--trunc_te 8 --freeze_te 1 \
	--train_batch_size 8 --nNeg 5 --seq_length 5 \
	--optim adam --lr 0.00001 --margin 0.1


Performances are below:

<img width="1136" height="81" alt="Image" src="https://github.com/user-attachments/assets/fa551583-fd7f-4b4f-9c47-d5dfd8277cce" />

**Comment by phatli** [comment]
From the args I can't tell if you freeze the correct layers of the backbone. As in the paper, I initiate with weight from CricaVPR and only train the last-layer adaptor in the backbone and the casenet.

**Comment by jinxxo-j** [comment]
Codes are below:

```python
if args.arch == "hvpr":
    model = HVPR_CaseNet(seq_len=args.seq_length, encoder_type="bs_d_c", is_pure=False)
    # Load DINOv2 weights from CricaVPR.pth
    cricavpr_checkpoint = torch.load("/home/jinwoo/vg-transformers/CricaVPR.pth", map_location='cpu')
    
    # Extract DINOv2 backbone weights from model_state_dict
    model_state = cricavpr_checkpoint['model_state_dict']
    dinov2_state_dict = {}
    
    for key, value in model_state.items():
        
        if key.startswith('module.backbone.'):
          
            new_key = key.replace('module.backbone.', '')
            dinov2_state_dict[new_key] = value
    

    missing_keys, unexpected_keys = model.encoder.load_state_dict(dinov2_state_dict, strict=False)
    
    if missing_keys:
        logging.info(f"Missing keys when loading DINOv2 weights: {len(missing_keys)} keys missing")
        logging.debug(f"Missing keys: {missing_keys[:10]}...")  
    if unexpected_keys:
        logging.info(f"Unexpected keys when loading DINOv2 weights: {len(unexpected_keys)} keys unexpected")
        logging.debug(f"Unexpected keys: {unexpected_keys[:10]}...")  
    
    logging.info(f"DINOv2 weights loaded from CricaVPR.pth - {len(dinov2_state_dict)} keys processed")
    

    # Freeze all parameters first
    for param in model.encoder.parameters():
        param.requires_grad = False
    
    # Find the last layer with adapter and enable only that layer's adapter
    adapter_layers = []
    for name, param in model.encoder.named_parameters():
        if "adapter" in name:
            # Extract layer number from parameter name
            layer_parts = name.split('.')
            for part in layer_parts:
                if part.isdigit():
                    layer_num = int(part)
                    adapter_layers.append((layer_num, name, param))
                    break
    
    if adapter_layers:
        # Sort by layer number and get the last (highest) layer
        adapter_layers.sort(key=lambda x: x[0])
        last_layer_num = adapter_layers[-1][0]
        
        # Enable only adapters in the last layer
        for layer_num, name, param in adapter_layers:
            if layer_num == last_layer_num:
                param.requires_grad = True
                logging.info(f"Enabling adapter parameter: {name}")
        
        logging.info(f"Only training adapters in the last layer (layer {last_layer_num})")
    else:
        logging.warning("No adapter parameters found in the encoder")
    
    # Set features_dim for HVPR model

    args.features_dim = 14 * 768


class HVPR_CaseNet(nn.Module):
    def __init__(self, seq_len=5, encoder_type="bs_d_c", is_pure=False):
        super().__init__()
        if is_pure:
            self.encoder = get_pure_dinov2(num_unfrozen_blocks=2)
        else:
            self.encoder = vit_base(patch_size=14, img_size=518,
                             init_values=1, block_chunks=0)
        self.pool = CaseNet(seq_len=seq_len, encode_type=encoder_type)
        self.seq_length = seq_len

        self.meta = {'outputdim': 14 * 768}
        
    def forward(self, x, single_img=False):
        features = self.encoder(x, is_training=True)
        seq_desc = self.pool(features, output_img=False, single_img=single_img)
        return  seq_desc
``` 

Using the code above, training runs for only five epochs and then stops because Recall@5 shows no improvement (early stopping). Is this expected behavior? Also, could you confirm approximately how many epochs of training the reported performance corresponds to?

I have another question.
CaseVPR322 is initialized using CricaVPR weights pre-trained on GSV-cities with 322 × 322 input, which employs the VPR-specific pre-trained weights [21: BoQ] of DINOv2 and follows the same training protocol as in [6: CricaVPR].:  

I’d like to confirm whether my understanding is correct. In the paper, CaseVPR322 is also initialized from CricaVPR’s weights. Does this mean that the CricaVPR weights were themselves initialized from BoQ’s DINOv2 weights and trained at 322×322, or does it simply mean that VPR-specific training on the GSV-Cities dataset was performed starting from standard DINOv2 weights?

**Comment by phatli** [comment]
The checkpoint I uploaded was at 14-th epoch. Maybe you stop too early. Given CaseNet encodes batches (b) and image sequences (s) as sequences to the transformer encoder, you should set the batch size to 1 for validation during training for better indication. You may also consider using Recall@1 as indicator to pick the final checkpoints as well. Beyond this, you may also play around with learning rate scheduler to gradually reduce the learning rate for finer convergence in later epochs.

For CaseVPR322, we load the backbone from BoQ's publicly available DINOv2 weights (as it shows strongest feature expressiveness) to initialize CricaVPR322, and follow the same pipeline in CricaVPR to adapt the model on CSV-cities dataset for cross-image correlation-aware feature learning. Later this trained CricaVPR322 is initialized to fine-tune CaseVPR322 on Oxford datasets for sequence-level representation learning.

Hope this clarify.
